### ISSUE TITLE
关于ctr_demo，特征维数过多（60000维）模型启动不起来的问题，有什么解决方案吗

### ISSUE THREAD

**Issue by irukawang** [PaddlePaddle/Paddle#16356]
为使您的问题得到快速解决，在建立Issue前，请您先通过如下方式搜索是否有相似问题:【搜索issue关键字】【使用labels筛选】【官方文档】

建立issue时，为快速解决问题，请您根据使用情况给出如下信息：
- 标题：简洁、精准描述您的问题，例如“ssd 模型前置lstm报错  ”
- 版本、环境信息：
    1）PaddlePaddle版本：请提供PaddlePaddle版本号，例如1.1或CommitID
    2）CPU：请提供CPU型号，MKL/OpenBlas/MKLDNN/等数学库的使用情况
    3）GPU：请提供GPU型号，CUDA和CUDNN版本号
    4）系统环境：请说明系统类型、版本（例如Mac OS 10.14），Python版本
- 模型信息
    1）模型名称 2）使用数据集名称 3）使用算法名称 4）模型链接
- 复现信息：如为报错，请给出复现环境、复现步骤
- 问题描述：请详细描述您的问题，同步贴出报错信息、日志/代码关键片段

Thank you for contributing to PaddlePaddle.
Before submitting the issue, you could search issue in the github.Probably there was a similar issue submitted or resolved before.
If there is no solution,please make sure that this is a issue of models including the following details:
**System information**
-PaddlePaddle version （eg.1.1）or CommitID
-CPU: including CPUMKL/OpenBlas/MKLDNN version
-GPU: including CUDA/CUDNN version
-OS Platform (eg.Mac OS 10.14)
-Python version
-Name of Models&Dataset/details of operator
**To Reproduce**
Steps to reproduce the behavior
**Describe your current behavior**
**Code to reproduce the issue**
**Other info / logs**

**Comment by heavengate** [comment]
能详细描述一下问题么，paddle版本，机器型号，训练脚本，参数配置等

**Comment by irukawang** [comment]
ctr_demo的例子里 特征就二十多维，  我们的是60000维的稀疏特征，模型运行不起来，一直卡在 optimizer.minimize(loss)地方。 版本是1.3

训练脚本：
```python
import paddle.fluid as fluid
from layers import embed, fc
import math

def ctr_dnn_model(py_reader, embed_size, sparse_feature_dim, is_training):
    """
    DNN模型
    :param py_reader: 输入py_reader
    :param embed_size: 词嵌入矩阵大小
    :param sparse_feature_dim: 稀疏矩阵维度
    :param is_training: 是否训练
    """
    words = fluid.layers.read_file(py_reader)
    sparse_feature, label = words[:sparse_feature_dim], words[-1]

    sparse_embed_seq = list(map(embed, sparse_feature))
    concated = fluid.layers.concat(sparse_embed_seq, axis=1)

    fc1 = fc(concated, size=400, act='relu', is_training=is_training)
    fc2 = fc(fc1, size=400, act='relu', is_training=is_training)
    fc3 = fc(fc2, size=400, act='relu', is_training=is_training)
    predict = fluid.layers.fc(fc3, size=2, act='softmax')

    return predict, label


def ctr_deepfm_model(factor_size, sparse_feature_dim, dense_feature_dim, sparse_input):
    """ctr deepfm model"""
    def dense_fm_layer(input, emb_dict_size, factor_size, fm_param_attr):
        """
        dense_fm_layer
        """
        first_order = fluid.layers.fc(input=input, size=1)
        emb_table = fluid.layers.create_parameter(shape=[emb_dict_size, factor_size],
                                                  dtype='float32', attr=fm_param_attr)

        input_mul_factor = fluid.layers.matmul(input, emb_table)
        input_mul_factor_square = fluid.layers.square(input_mul_factor)
        input_square = fluid.layers.square(input)
        factor_square = fluid.layers.square(emb_table)
        input_square_mul_factor_square = fluid.layers.matmul(input_square, factor_square)

        second_order = 0.5 * (input_mul_factor_square - input_square_mul_factor_square)
        return first_order, second_order

    def sparse_fm_layer(input, emb_dict_size, factor_size, fm_param_attr):
        """
        sparse_fm_layer
        """
        first_embeddings = fluid.layers.embedding(
            input=input, dtype='float32', size=[emb_dict_size, 1], is_sparse=True)
        first_order = fluid.layers.sequence_pool(input=first_embeddings, pool_type='sum')

        nonzero_embeddings = fluid.layers.embedding(
            input=input, dtype='float32', size=[emb_dict_size, factor_size],
            param_attr=fm_param_attr, is_sparse=True)
        summed_features_emb = fluid.layers.sequence_pool(input=nonzero_embeddings, pool_type='sum')
        summed_features_emb_square = fluid.layers.square(summed_features_emb)

        squared_features_emb = fluid.layers.square(nonzero_embeddings)
        squared_sum_features_emb = fluid.layers.sequence_pool(
            input=squared_features_emb, pool_type='sum')

        second_order = 0.5 * (summed_features_emb_square - squared_sum_features_emb)
        return first_order, second_order

    dense_input = fluid.layers.data(name="dense_input", shape=[dense_feature_dim], dtype='float32')

    sparse_input_ids = [
        fluid.layers.data(name="C" + str(i), shape=[1], lod_level=1, dtype='int64')
        for i in range(1, 27)]

    label = fluid.layers.data(name='label', shape=[1], dtype='int64')

    datas = [dense_input] + sparse_input_ids + [label]

    py_reader = fluid.layers.create_py_reader_by_data(capacity=64,
                                                      feed_list=datas,
                                                      name='py_reader',
                                                      use_double_buffer=True)
    words = fluid.layers.read_file(py_reader)

    sparse_fm_param_attr = fluid.param_attr.ParamAttr(name="SparseFeatFactors",
                                                      initializer=fluid.initializer.Normal(
                                                          scale=1 / math.sqrt(sparse_feature_dim)))
    dense_fm_param_attr = fluid.param_attr.ParamAttr(name="DenseFeatFactors",
                                                     initializer=fluid.initializer.Normal(
                                                         scale=1 / math.sqrt(dense_feature_dim)))

    sparse_fm_first, sparse_fm_second = sparse_fm_layer(
        sparse_input, sparse_feature_dim, factor_size, sparse_fm_param_attr)
    dense_fm_first, dense_fm_second = dense_fm_layer(
        dense_input, dense_feature_dim, factor_size, dense_fm_param_attr)

    def embedding_layer(input):
        """embedding_layer"""
        emb = fluid.layers.embedding(
            input=input, dtype='float32', size=[sparse_feature_dim, factor_size],
            param_attr=sparse_fm_param_attr, is_sparse=True)
        return fluid.layers.sequence_pool(input=emb, pool_type='average')

    sparse_embed_seq = list(map(embedding_layer, sparse_input_ids))
    concated = fluid.layers.concat(sparse_embed_seq + [dense_input], axis=1)
    fc1 = fluid.layers.fc(input=concated, size=400, act='relu',
                          param_attr=fluid.ParamAttr(initializer=fluid.initializer.Normal(
                              scale=1 / math.sqrt(concated.shape[1]))))
    fc2 = fluid.layers.fc(input=fc1, size=400, act='relu',
                          param_attr=fluid.ParamAttr(initializer=fluid.initializer.Normal(
                              scale=1 / math.sqrt(fc1.shape[1]))))
    fc3 = fluid.layers.fc(input=fc2, size=400, act='relu',
                          param_attr=fluid.ParamAttr(initializer=fluid.initializer.Normal(
                              scale=1 / math.sqrt(fc2.shape[1]))))
    predict = fluid.layers.fc(
        input=[sparse_fm_first, sparse_fm_second, dense_fm_first, dense_fm_second, fc3],
        size=2,
        act="softmax",
        param_attr=fluid.ParamAttr(initializer=fluid.initializer.Normal(scale=1 / math.sqrt(fc3.shape[1]))))

    cost = fluid.layers.cross_entropy(input=predict, label=words[-1])
    avg_cost = fluid.layers.reduce_sum(cost)
    accuracy = fluid.layers.accuracy(input=predict, label=words[-1])
    auc_var, batch_auc_var, auc_states = \
        fluid.layers.auc(input=predict, label=words[-1], num_thresholds=2 ** 12, slide_steps=20)

    return avg_cost, auc_var, batch_auc_var, py_reader
```

```python
import paddle
import paddle.fluid as fluid
from layers import embed, fc, cross_entropy, acc, auc
from module import ctr_dnn_model
from reader import Reader
from config import Config
from multiprocessing import cpu_count
import numpy as np
import math
import logging
import os
import time

logging.basicConfig(format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger("fluid")
logger.setLevel(logging.INFO)

class ctr_model(object):
    def __init__(self, args):
        """
        创建整个模型
        :param args: 配置文件
        """
        self.args = args
        self._create_placeholders()
        self._create_model()
        self._create_losses()
        self._create_matrics()
        self._create_opts()

    def _create_placeholders(self):
        """
        创建占位符
        """
        logger.info("create placeholders")

        self.sparse_feature_ph = [fluid.layers.data(name="sparse_feature_ph_%d" % idx, shape=[1], dtype="int64")
                                for idx in xrange(self.args.feature_dim)]
        self.label_ph = fluid.layers.data(name="label", shape=[1], dtype="int64")

        self.py_reader = fluid.layers.create_py_reader_by_data(
                            capacity=64,
                            feed_list=self.sparse_feature_ph + [self.label_ph],
                            use_double_buffer=True,
                            name="py_reader")

    def _create_model(self):
        """
        创建网络
        """
        logger.info("create networks")

        self.predict, self.label = ctr_dnn_model(self.py_reader, self.args.embed_size, self.args.sparse_feature_dim, self.args.is_training)

    def _create_losses(self):
        """
        创建损失函数
        """
        logger.info("create loss function")
        self.loss = cross_entropy(self.predict, self.label)

    def _create_matrics(self):
        """
        创建模型评价指标
        """
        logger.info("create matrics")
        self.acc = acc(self.predict, self.label)
        self.auc, self.batch_auc, self.auc_states = auc(self.predict, self.label)

    def _create_opts(self):
        """
        创建优化器
        """
        logger.info("create optimizer")
        self.optimizer = fluid.optimizer.Adam(learning_rate=self.args.lr)
        self.optimizer.minimize(self.loss)

    def train(self):
        """
        循环迭代
        """
        logger.info("run local training")
        if not os.path.exists(self.args.model_output_dir):
            os.makedirs(self.args.model_output_dir)

        dataset = Reader(self.args.sparse_feature_dim, [self.args.train_dir]).feed()
        train_reader = paddle.batch(paddle.reader.shuffle(dataset, buf_size=self.args.batch_size * 100),
                                    batch_size=self.args.batch_size)

        self.py_reader.decorate_paddle_reader(train_reader)
        data_name_list = []

        place = fluid.CPUPlace()
        exe = fluid.Executor(place)

        exec_strategy = fluid.ExecutionStrategy()
        build_strategy = fluid.BuildStrategy()

        if os.getenv("NUM_THREADS", ""):
            exec_strategy.num_threads = int(os.getenv("NUM_THREADS"))

        cpu_num = int(os.environ.get('CPU_NUM', cpu_count()))
        build_strategy.reduce_strategy = \
                fluid.BuildStrategy.ReduceStrategy.Reduce if cpu_num > 1 \
                else fluid.BuildStrategy.ReduceStrategy.AllReduce

        self.main_program = fluid.default_main_program()
        pe = fluid.ParallelExecutor(
            use_cuda=False,
            loss_name=self.loss.name,
            main_program=self.main_program,
            build_strategy=build_strategy,
            exec_strategy=exec_strategy)

        exe.run(fluid.default_startup_program())

        for pass_id in xrange(self.args.num_passes):
            pass_start = time.time()
            batch_id = 0
            self.py_reader.start()

            try:
                while True:
                    loss, acc, auc, batch_auc = pe.run(fetch_list=[self.loss.name, self.acc.name, self.auc.name, self.batch_auc.name])
                    loss = np.mean(loss)
                    auc = np.mean(auc)
                    batch_auc = np.mean(batch_auc)
                    logger.info("TRAIN --> pass: {} batch: {} loss: {} acc: {}  auc: {}, batch_auc: {}"
                            .format(pass_id, batch_id, loss / self.args.batch_size, acc,  auc, batch_auc))
                    if batch_id % 1000 == 0 and batch_id != 0:
                        model_dir = args.model_output_dir + '/batch-' + str(batch_id)
                        if args.trainer_id == 0:
                            fluid.io.save_inference_model(model_dir, data_name_list, [self.loss, self.auc], exe)
                    batch_id += 1

            except fluid.core.EOFException:
                self.py_reader.reset()

            print("pass_id: %d, pass_time_cost: %f" % (pass_id, time.time() - pass_start))

            model_dir = args.model_output_dir + '/pass-' + str(pass_id)
            if args.trainer_id == 0:
                fluid.io.save_inference_model(model_dir, data_name_list, [self.loss, self.auc], exe)


if __name__ == "__main__":
    args = Config()
    args = args()

    ctr_demo = ctr_model(args)
    ctr_demo.train()
```
